src/feat/signal-distort.cc

```diff
@@ -72,5 +72,90 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
   perturb_egs->CopyFromMat(out_mat);
 }
 
+// This function add the noise to the orginial signal. We should not normalize 
+// the signal level of the orginial signal. According to SNR, we rescale the noise
+// and add it. So that the perturbed signal is created. 
+void PerturbXvectorSignal::ApplyAdditiveNoise(const MatrixBase<BaseFloat> &input_eg,
+                                              const Matrix<BaseFloat> &noise_mat,
+                                              Matrix<BaseFloat> *perturbed_eg) {
+  // In the version, we ask the noise_cols == input_cols.
+  int32 input_rows = input_eg.NumRows(), input_cols = input_eg.NumCols();  
+  KALDI_ASSERT(noise_mat.NumCols() == input_cols);
+
+  // As the noise_mat is very huge and the input_eg is small normally,
+  // so we'd better not reload the "noise_mat" matrix
+  // select the noise range
+
+  Matrix<BaseFloat> selected_noise_mat;
+  selected_noise_mat.Resize(input_rows, input_cols);
+  
+  int32 noise_rows = noise_mat.NumRows();
+  int32 start_row_ind = RandInt(0, noise_rows - input_rows);
+  
+  if (noise_mat.NumRows() < input_rows) {
+    int32 indices[input_rows];
+    for (int32 i=0; i < input_rows; ++i) {
+      indices[i] = (start_row_ind + i) % noise_mat.NumRows();
+    }
+    selected_noise_mat.CopyRows(noise_mat, indices);
+  } else {
+    selected_noise_mat.AddMat(1.0, noise_mat.Range(start_row_ind, input_rows,
+                                                  0, input_cols));
+  }
+
+  // compute the energy of noise and input
+  Matrix<BaseFloat> input_energy_mat(input_rows, input_cols);
+  input_energy_mat.AddMatMatElements(1.0, input_eg, input_eg, 0.0);
+  double input_energy = input_energy_mat.Sum();
+  Matrix<BaseFloat> noise_energy_mat(input_rows, input_cols);
+  noise_energy_mat.AddMatMatElements(1.0, selected_noise_mat, selected_noise_mat, 0.0);
+  double noise_energy = noise_energy_mat.Sum();
+
+  // In Energy domain, SNR=20log10(S/N). 
+  // 10^(SNR/20) = input_energy / (scale^2 * noise_energy)
+  double scale = input_energy / noise_energy / (pow(10,opts_.snr/20));
+  scale = sqrt(scale);
+  
+  // Add noise mat to input_eg mat
+  perturbed_eg->Resize(input_rows, input_cols);
+  perturbed_eg->CopyFromMat(input_eg);
+  perturbed_eg->AddMat(scale, selected_noise_mat);
+}
+
+void PerturbXvectorSignal::ApplyDistortion(const MatrixBase<BaseFloat> &input_egs,
+                                           Matrix<BaseFloat> *perturb_egs) {
+  if (!opts_.add_noise.empty()) {
+    // choose a noise from the noise.scp/ark
+    // 1) we need to record the keys of noise_egs
+    std::vector<std::string> noise_list;
+    SequentialBaseFloatMatrixReader noise_seq_reader(opts_.add_noise);
+    for (; !noise_seq_reader.Done(); noise_seq_reader.Next()) {
+      std::string key = noise_seq_reader.Key();
+      noise_list.push_back(key);
+    }
+    noise_seq_reader.Close();
```

Comment:
This part is not efficient, since you need to generate list of noises, each time you call this function for each example.
The best thing is that you add private variable std::vector<std::string> noise_lists_   to PertubXvectorSignal and then initialize that in class constructor and then you can randomly pick a noise from noise list in ApplyDistortion.

Also PerturbExample should accept object of PerturbXvectorSignal as input as PerturbExample(PertubXvectorSignal &example_perturber, Matrix<BaseFloat> &eg, Matix<BaseFloat> *perturbed_eg) and you should generate object of PertubXvectorSignal class before example loop.

Another alternative is to remove ApplyDistortion function from  PertubXvectorSignal class and Do all those steps in PertubExample and then make AddAdditiveNoise as public function in Xvector class.